src/util/persistent_logger.js

```diff
@@ -264,6 +247,78 @@ class LogFile {
     constructor(fs_context, log_path) {
         this.fs_context = fs_context;
         this.log_path = log_path;
+        this.log_reader = new NewlineReader(
+            this.fs_context,
+            this.log_path, { lock: 'EXCLUSIVE', skip_overflow_lines: true, skip_leftover_line: true },
+        );
+    }
+
+    /**
+     * init eagerly initializes the underlying log reader which also means
+     * that it will immediately acquire an EX lock on the underlying file.
+     * 
+     * Calling this method isn't necessary as both `collect` and `collect_and_process`
+     * lazily initializes the reader. However, explicit initializing can help
+     * manage the lifecycle of the underlying lock.
+     */
+    async init() {
+        if (this.log_reader.fh) return;
+        await this.log_reader.init();
+    }
+
+    /**
+     * deinit closes the underlying log file and will reset the reader as well.
+     * This also means that deinit will lose the lock hence this should be called
+     * with great care.
+     */
+    async deinit() {
+        if (!this.log_reader.fh) return;
+        await this.log_reader.close();
+        this.log_reader.reset();
```

Comment:
Why reset here? We have reset at both `collect` and `collect_and_process` right before we start processing.